spring-integration-azure/spring-integration-servicebus/src/main/java/com/microsoft/azure/spring/integration/servicebus/queue/ServiceBusQueueTemplate.java

```diff
@@ -142,19 +144,22 @@ protected String buildCheckpointSuccessMessage(Message<?> message) {
         // ISessionHandler
         @Override
         public CompletableFuture<Void> onMessageAsync(IMessageSession session, IMessage serviceBusMessage) {
-            
             Map<String, Object> headers = new HashMap<>();
-            
-            Message<U> message = messageConverter.toMessage(serviceBusMessage, new MessageHeaders(headers),
-                    payloadType);
-            consumer.accept(message);
-            
-            try {
-                session.complete(serviceBusMessage.getLockToken());
-            } catch (InterruptedException | ServiceBusException e) {
-                this.notifyException(e, ExceptionPhase.COMPLETE);
+
+            Checkpointer checkpointer = new AzureCheckpointer(
+                    () -> session.completeAsync(serviceBusMessage.getLockToken()),
+                    () -> session.abandonAsync(serviceBusMessage.getLockToken()));
+            if (checkpointConfig.getCheckpointMode() == CheckpointMode.MANUAL) {
+                headers.put(AzureHeaders.CHECKPOINTER, checkpointer);
             }
 
+            Message<U> message = messageConverter.toMessage(serviceBusMessage,
+                    new MessageHeaders(headers), payloadType);
+            consumer.accept(message);
+
+            if (checkpointConfig.getCheckpointMode() == CheckpointMode.RECORD) {
+                return checkpointer.success().whenComplete((v, t) -> super.checkpointHandler(message, t));
+            }
             return CompletableFuture.completedFuture(null);
         }
```

Comment:
Any chance we have unit tests for this changes?

Reply:
yep, I will add unit tests for it later.